core/auth_provider_okta.go

```diff
@@ -0,0 +1,311 @@
+
+	// "github.com/jhunt/go-log"
+	"github.com/shieldproject/shield/db"
+	"github.com/shieldproject/shield/route"
+	"github.com/shieldproject/shield/util"
+	
+	"github.com/pborman/uuid"
+	"github.com/thanhpk/randstr"
+
+	verifier "github.com/okta/okta-jwt-verifier-golang"
+)
+
+type OktaAuthProvider struct {
+	AuthProviderBase
+
+	ClientID         	string `json:"client_id"`
+	ClientSecret     	string `json:"client_secret"`
+	AuthorizationServer string `json:"authorization_server"`
+	OktaDomain		 	string `json:"okta_domain"`
+	DeploymentURI		string `json:"deployment_uri"`
+
+	OktaEnterprise 	 bool   `json:"okta_enterprise"`
+	Mapping          []struct {
+		Okta string `json:"okta"`
+		Tenant string `json:"tenant"`
+		Rights []struct {
+			Group string `json:"group"`
+			Role string `json:"role"`
+		} `json:"rights"`
+	} `json:"mapping"`
+}
+
+type Exchange struct {
+	Error            string `json:"error,omitempty"`
+	ErrorDescription string `json:"error_description,omitempty"`
+	AccessToken      string `json:"access_token,omitempty"`
+	TokenType        string `json:"token_type,omitempty"`
+	ExpiresIn        int    `json:"expires_in,omitempty"`
+	Scope            string `json:"scope,omitempty"`
+	IdToken          string `json:"id_token,omitempty"`
+}
+
+func (p *OktaAuthProvider) Configure(raw map[interface{}]interface{}) error {
+	b, err := json.Marshal(util.StringifyKeys(raw))
+	if err != nil {
+		return err
+	}
+
+	err = json.Unmarshal(b, p)
+	if err != nil {
+		return err
+	}
+
+	if p.ClientID == "" {
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `client_id' value")
+	}
+
+	if p.ClientSecret == "" {
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `client_secret' value")
+	}
+
+	if p.OktaDomain == "" {
+		// ex: p.OktaDomain = "https://dev-xyz.okta.com/" 
+		// p.OktaIssuer = "https://dev-xyz.okta.com/oauth2/default" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `okta_domain' value")
+	}
+
+	if p.DeploymentURI == "" {
+		// ex: p.DeploymentURI = "https://shield.starkandwayne.com" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `deployment_uri' value")
+	}
+
+	if p.AuthorizationServer == "" {
+		// ex: p.AuthorizationServer = "default" 
+		return fmt.Errorf("invalid configuration for Okta OAuth Provider: missing `authorization_server' value")
+	}
+
+
+	p.OktaDomain = strings.TrimSuffix(p.OktaDomain, "/")
+	p.DeploymentURI = strings.TrimSuffix(p.DeploymentURI, "/")
+
+
+	p.properties = util.StringifyKeys(raw).(map[string]interface{})
+
+	return nil
+}
+
+func (p *OktaAuthProvider) WireUpTo(c *Core) {
+	p.core = c
+}
+
+func (p *OktaAuthProvider) ReferencedTenants() []string {
+	ll := make([]string, 0)
+	for _, m := range p.Mapping {
+		ll = append(ll, m.Tenant)
+	}
+	return ll
+}
+
+func (p *OktaAuthProvider) Initiate(r *route.Request) {
+
+	var state = randstr.Hex(16)
+
+	var redirectPath string
+
+	q := r.Req.URL.Query()
```

Comment:
Is there a reason why we us the request `URL.Query` instead of a new instance of Values? https://pkg.go.dev/net/url#Values

Reply:
url.Values{} still needs to add each parameter in a single line. I can maybe make a custom struct for it?

Reply:
You can simply do like the following:
```golang
...
	q := url.Values{}
	q.Add("client_id", p.ClientID)
	q.Add("response_type", "code")
...
```